File: cpp/n101_IsSymmetric.cpp
// https://leetcode-cn.com/problems/symmetric-tree/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
bool isReverse(vector<int> &arr){
    if (arr.empty()) return true;
    int i = 0;
    int j = arr.size() - 1;
    while(j > i){
        if (arr[i] != arr[j]) return false;
        ++i;
        --j;
    }
    return true;
} 
 
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        vector<int> valueQueue = {};
        if (root != NULL) valueQueue.push_back(root -> val);
        vector<TreeNode*> nodeQueue = {root};
        while(!nodeQueue.empty()){
            int length = nodeQueue.size();
            for (int i=0; i < length; ++i){
                TreeNode* node = nodeQueue.front();                             
                if (!valueQueue.empty()) {valueQueue.erase(valueQueue.begin());}
                if (node != NULL){
                    nodeQueue.push_back(node -> left);
                    cout << node->val << endl;
                    if (node -> left != NULL) {valueQueue.push_back(node -> left -> val);}
                    else valueQueue.push_back(-1);
                    nodeQueue.push_back(node -> right);
                    if (node -> right != NULL) valueQueue.push_back(node -> right -> val);
                    else valueQueue.push_back(-1);                                     
                }
                nodeQueue.erase(nodeQueue.begin());
            }
            if (!isReverse(valueQueue)) return false;
        }
        return true;        
    }   
};

int main(){
    vector<int> arr = {3, NULL, NULL, 3};
    cout << isReverse(arr) << endl;
}

File: python/n101_IsSymmetric.py
# https://leetcode-cn.com/problems/symmetric-tree/
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def isReverse(arr: list) -> bool:
    return arr[::-1] == arr

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        valueQueue = []
        if root: valueQueue = [root.val]
        nodeQueue = [root]
        while nodeQueue:
            for i in range(len(nodeQueue)):
                node = nodeQueue.pop(0)
                if valueQueue: valueQueue.pop(0)
                if node:
                    nodeQueue.append(node.left)
                    if node.left: valueQueue.append(node.left.val)
                    else: valueQueue.append(None)
                    nodeQueue.append(node.right)
                    if node.right: valueQueue.append(node.right.val)
                    else: valueQueue.append(None)
            if not isReverse(valueQueue):
                return False
        return True  
        
        
if __name__ == "__main__":
    arr = [1, None, 1, None]
    print(arr.pop(0))
    print(arr)